### PR TITLE
Implement secondary and caching registry

### LINKED ISSUE #2449: Registry: Implement secondary zones
A registry should be able to be non-authoritative, however we need to see how this is best implemented, especially taking https://github.com/bosagora/agora/pull/2443 into account.

### LINKED ISSUE #2647: DNS: Implement whitelisting for AXFR
We should have a way to restrict AXFR to whitelisted clients, so that it doesn't get abused.
A config file entry, probably at the registry level, as it doesn't make much sense to have it at the zone level, would be needed.

```yaml
registry:
  enabled: true
  whitelist:
    - 1.2.3.4
    - foo.bar
```

### PR DESCRIPTION
**Pull request by mkykadir**
This PR implements secondary and caching registry. Fixes of #2449 and Fixes #2647
With secondary registries we can create backup registry servers to improve availability, also this implementation enables us to create a hidden registry.

- It uses `DNSResolver`, depends on #2636 
- Requests AXFR over UDP
- SOA RR is checked periodically, depends on #2730 
- POST endpoints are redirected to actual primary
- Removes two `boolean` flags from configuration and creates three type registries
  - primary registry server, only this will have database modifying capabilities, authoritative
  - secondary registry server, this will do AXFR query to its primaries; secondary servers can be primary for other secondaries
  - caching server, not implemented yet but will be non-authoritative
- Previous Registry cache is removed when Zone type is updated
- Caching zone starts with clean storage and adds to cache according to query results up to TTL seconds of the record.
- Makes Registry enabled by default with caching zones

**DNSResolver**

- [ ] RFC recommends TCP for AXFR, currently using UDP

### CODE REVIEW

**Comment by mkykadir** [comment]
@Geod24 this is ready for review (even it depends on WIP Resolver :) )

File: source/agora/common/Types.d [hunk 1]
```diff
@@ -34,6 +34,7 @@ public T clone (T)(in T input)
 shared static this ()
 {
     registerCommonInternetSchema("agora", 2826);
+    registerCommonInternetSchema("dns", 53);
```
Comment (Geod24) [inline source/agora/common/Types.d]: I've included this commit in my rebase

File: source/agora/node/Registry.d [hunk 2]
```diff
@@ -751,14 +753,28 @@ private struct ZoneData
             "FOREIGN KEY(pubkey) REFERENCES registry_%s_utxo(pubkey) ON DELETE CASCADE, " ~
             "PRIMARY KEY(pubkey, address))", zone_name, zone_name);
 
+        // Initialize common fields
+        this.soa = this.config.fromConfig(this.root);
+
         this.db.execute(query_sig_create);
         this.db.execute(query_addr_create);
+        this.updateSOA();
```
Comment (Geod24) [inline source/agora/node/Registry.d]: I think this call should be right after `this.soa = ...` (L757).

File: source/agora/node/Registry.d [hunk 3]
```diff
@@ -525,12 +527,12 @@ unittest
...
 
 /// Converts a `ZoneConfig` to an `SOA` record
-private SOA fromConfig (in ZoneConfig zone, Domain name, uint serial) @safe pure
+private SOA fromConfig (in ZoneConfig zone, Domain name) @safe pure
 {
     SOA soa;
     soa.mname = Domain(format("ns1.%s", name.value));
     soa.rname = Domain(zone.soa.email.value.replace('@', '.'));
-    soa.serial = serial;
+    soa.serial = 0;
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Just rm the line

File: devel/config-fullnode.yaml [hunk 4]
```diff
@@ -33,9 +33,13 @@ validator:
...
   flash:
-    email: github@bosagora.io
+    type: primary
+    soa:
+      email: github@bosagora.io
   validators:
-    email: github@bosagora.io
+    type: primary
+    soa:
+      email: github@bosagora.io
```
Comment (Geod24) [inline devel/config-fullnode.yaml]: The commit message should explain why this is done.
I'm reviewing commit-per-commit and currently I can't tell why we should do this. If an `SOA` is specified, surely it's a `primary` zone. If it's `authoritative` but no `SOA` is specified, surely it's a secondary zone ?

**Comment by mkykadir** [comment]
Rebased

File: source/agora/node/Registry.d [hunk 5]
```diff
@@ -737,14 +741,21 @@ private struct ZoneData
         this.config = config;
         this.root = root;
 
-        static string serverType (bool auth, bool primary)
+        static string serverType (ZoneConfig.Type zone_type)
         {
-            return primary ? "primary" : (auth ? "secondary" : "caching");
+            switch (zone_type)
+            {
+                case ZoneConfig.Type.disabled: return "disabled";
```
Comment (Geod24) [inline source/agora/node/Registry.d]: The idea is that we never disable it.

File: source/agora/node/Registry.d [hunk 6]
```diff
@@ -129,8 +136,8 @@ public class NameRegistry: NameRegistryAPI
     private void collectStats (Collector collector)
     {
         RegistryStats stats;
-        stats.registry_validator_record_count = this.zones[this.validators].count();
-        stats.registry_flash_record_count = this.zones[this.flash].count();
+        stats.registry_validator_record_count = this.zones[1].count();
+        stats.registry_flash_record_count = this.zones[2].count();
```
Comment (Geod24) [inline source/agora/node/Registry.d]: I wouldn't mind this refactoring, but in its own commit. Also we should use an `enum` here.

File: source/agora/node/Runner.d [hunk 7]
```diff
@@ -320,7 +320,7 @@ private void runDNSServer_canThrow (in RegistryConfig config, NameRegistry regis
             auto pack = udp.recv(buffer, ppeer);
             auto query = deserializeFull!Message(pack);
             registry.answerQuestions(
-                query,
+                query, peer.toAddressString,
```
Comment (Geod24) [inline source/agora/node/Runner.d]: Does it do a reverse DNS lookup or just spits out an IP ?

File: source/agora/node/Registry.d [hunk 8]
```diff
@@ -351,8 +351,8 @@ public class NameRegistry: NameRegistryAPI
     {
         Message reply;
         reply.header.RCODE = Header.RCode.FormatError;
-        reply.header.RA = false; // TODO: Implement
-        reply.header.AA = true;  // TODO: Make configurable
+        reply.header.RA = false;
+        reply.header.AA = true;
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Caching shouldn't be AA should it ?

**Review by Geod24 — COMMENTED** [review]
Did a first pass.

File: source/agora/node/Registry.d (hunk 8 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: [This](https://github.com/bosagora/agora/pull/2703/files#diff-9a7d511b8407367630fb104c60937f8c304fb9d6eed6ff2281d46541c93513f9R1174) handles that. I will move setting those fields into `answer` method within the zone.

File: devel/config-fullnode.yaml (hunk 4 above)
Comment (mkykadir) [reply devel/config-fullnode.yaml]: It doesn't differ, I thought that this way would be more clearer to differentiate between different zone types (while configuring and coding) commit message states that but if it seems unnecessary to you, I can revert.

File: source/agora/node/Registry.d (hunk 6 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: This was required due to new additions to `ZoneConfig`; will be separated.

File: source/agora/node/Registry.d (hunk 3 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: That's outdated; #2788 changed this in a way that parameters are passed directly to Ctor

File: source/agora/node/Runner.d (hunk 7 above)
Comment (mkykadir) [reply source/agora/node/Runner.d]: Just string representation of the IP

**Comment by mkykadir** [comment]
Rebased, I needed to remove `unittest` of the Registry since it now requires a TaskManager to set-up timers. #2790 will open a door for adding proper Registry tests under `tests`

File: source/agora/node/Registry.d (hunk 6 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: Separated and merged in new PR #2825

File: devel/config-fullnode.yaml (hunk 4 above)
Comment (mkykadir) [reply devel/config-fullnode.yaml]: Reverted as requested

**Comment by mkykadir** [comment]
Rebased

File: tests/system/node/0/config.yaml [hunk 9]
```diff
@@ -48,6 +48,26 @@ validator:
...
+registry:
+  enabled: true
+  port: 3053
+
+  validators:
+    authoritative: true
+    primary: integration.testnet
+
+    allow_transfer:
+      - node-2
```
Comment (mkykadir) [inline tests/system/node/0/config.yaml]: This field requires IP address to be provided.

File: source/agora/test/Base.d [hunk 10]
```diff
@@ -2312,7 +2313,18 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
...
-      registry: { enabled: true, },
+      registry: {
+          validators: {
+              authoritative: true,
+              primary: "test.name.registry",
+              soa: { email: SetInfo!string("test@testnet"), },
+          },
+          flash: {
+              authoritative: true,
+              primary: "test.name.registry",
```
Comment (Geod24) [inline source/agora/test/Base.d]: Is that correct ? Shouldn't that be `name.registry`, as this one is the authoritative registry ?

File: source/agora/node/Registry.d [hunk 11]
```diff
@@ -576,13 +578,13 @@ unittest
...
 /// Converts a `ZoneConfig` to an `SOA` record
-private SOA fromConfig (in ZoneConfig zone, Domain name, uint serial) @safe
+private SOA fromConfig (in ZoneConfig zone, Domain name) @safe
 {
     return SOA(
         // mname, rname
         Domain.fromString(zone.primary),
         Domain.fromString(zone.soa.email.value.replace('@', '.')),
-        serial,
+        0,
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Add a comment saying it's set from `updateSOA` ? Or better, move `updateSOA` at the end of `fromConfig` ?

File: source/agora/node/Registry.d [hunk 12]
```diff
@@ -823,14 +825,28 @@ private struct ZoneData
...
+
+    public void updateSOA () @trusted
+    {
         if (this.type == ZoneType.primary)
         {
-            // Serial's value wraps around so the cast is safe
-            const serial = cast(uint) Clock.currTime(UTC()).toUnixTime();
-            this.soa = this.config.fromConfig(this.root, serial);
+            this.soa.serial = cast(uint) Clock.currTime(UTC()).toUnixTime();
+            return;
```
Comment (Geod24) [inline source/agora/node/Registry.d]:
```D
if (this.type != ZoneType.primary)
    return;
this.soa.serial = ...;
```

The `else` is not required anymore since we're no longer in the constructor.

File: doc/config.example.yaml [hunk 13]
```diff
@@ -363,6 +364,20 @@ registry:
 
   # Zone responsible for flash nodes registration
   flash:
-    # Having separate authoritative server increases redundancy and reduces load,
+    # Having a secondary server increases redundancy and reduces load,
```
Comment (Geod24) [inline doc/config.example.yaml]: "one or more secondary servers" ?

File: source/agora/node/Config.d [hunk 14]
```diff
@@ -463,9 +463,20 @@ public struct RegistryConfig
...
+            if (!zone.primary.set && zone.authoritative.set)
+            {
+                ensure(zone.primary_servers.length > 0,
+                    "registry.{} is secondary and at least one primary is required",
+                    name);
+
+                ensure(zone.redirect_primary.length > 0,
+                    "registry.{} is secondary (authoritative), thus requires an " ~
+                    "address registering endpoint to an Agora node",
+                    name);
```
Comment (Geod24) [inline source/agora/node/Config.d]: Perhaps we should make it default to `primary` ? It seems that in most configuration, that'd be the same.

File: source/agora/node/Config.d [hunk 15]
```diff
@@ -463,9 +463,20 @@ public struct RegistryConfig
...
                    "'email' fields to be set", name);
 
-            // If we're not primary, there's no configuration to validate
-            if (!zone.primary.set)
+            // Secondary registry
+            if (!zone.primary.set && zone.authoritative.set)
+            {
+                ensure(zone.primary_servers.length > 0,
+                    "registry.{} is secondary and at least one primary is required",
+                    name);
```
Comment (Geod24) [inline source/agora/node/Config.d]: Hold on... We have `primary`, `primary_servers`, and `redirect_primary` ? That's a bit much, isn't it ?

File: source/agora/node/Registry.d [hunk 16]
```diff
@@ -226,6 +232,12 @@ public class NameRegistry: NameRegistryAPI
 
     public override void postValidator (RegistryPayload registry_payload)
     {
+        if (this.zones[ZoneIndex.Validator].type == ZoneType.secondary)
+        {
+            this.zones[ZoneIndex.Validator].redirect_primary.postValidator(registry_payload);
+            return;
+        }
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Any redirect should happen after we've checked the payload is valid (at the very least the signature), otherwise it's a very obvious DDoS vector.

File: source/agora/node/Registry.d [hunk 17]
```diff
@@ -314,6 +332,7 @@ public class NameRegistry: NameRegistryAPI
         // register data
         log.info("Registering network addresses: {} for Flash public key: {}", registry_payload.data.addresses,
             registry_payload.data.public_key.toString());
+
```
Comment (Geod24) [inline source/agora/node/Registry.d]: ?

File: source/agora/node/Registry.d [hunk 18]
```diff
@@ -583,7 +617,9 @@ private SOA fromConfig (in ZoneConfig zone, Domain name) @safe
     return SOA(
         // mname, rname
         Domain.fromString(zone.primary),
-        Domain.fromString(zone.soa.email.value.replace('@', '.')),
+        (zone.soa.email.length > 0) ?
+            Domain.fromString(zone.soa.email.value.replace('@', '.'))
+            : Domain.init, // Secondary will get this from primary
```
Comment (Geod24) [inline source/agora/node/Registry.d]: It will get the rest from primary too, so maybe we should just not be calling `fromConfig` for non-primary ?

File: source/agora/node/Runner.d [hunk 19]
```diff
@@ -305,7 +305,8 @@ private void runDNSServer_canThrow (in RegistryConfig config, NameRegistry regis
     // The `listenUDP` needs to be in the `runTask` otherwise we get
     // a fatal error due to a bug in vibe-core (see comment #2):
     /// https://github.com/vibe-d/vibe-core/issues/289
-    auto udp = listenUDP(config.port, config.address);
+    auto udp = listenUDP(config.port, config.address,
+        UDPListenOptions.reuseAddress | UDPListenOptions.reusePort);
```
Comment (Geod24) [inline source/agora/node/Runner.d:309]: Own commit please

File: doc/config.example.yaml [hunk 20]
```diff
@@ -342,6 +342,14 @@ registry:
     allow_transfer:
       - 127.0.0.1
 
+    # Whitelist for AXFR zone transfers
+    #
+    # Only servers listed in this field will be allowed to initiate an AXFR zone
+    # transfer from this server. Server will not allow zone transfer to any
+    # requester if not configured. Applicable for primary and secondary servers.
+    allow_transfer:
+      - ns2.bosagora.io
```
Comment (Geod24) [inline doc/config.example.yaml]: Can't be a domain, can it ?

Comment (Geod24) [reply doc/config.example.yaml]: Also should be merged in the commit that implements it

**Review by Geod24 — CHANGES_REQUESTED** [review]
First two commits are G2G, need to do a bit more review, but left a few comments already.

File: source/agora/node/Config.d (hunk 15 above)
Comment (mkykadir) [reply source/agora/node/Config.d]: Yes, when I was started `primary` wasn't there, thus I need to think over that again. `primary_servers` can have multiple servers configured as a backup.

File: source/agora/node/Registry.d (hunk 18 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: Non-primaries call this to initialize timing fields of their SOA as a backup value (in-case of they couldn't reach any primary to get actual SOA)

File: doc/config.example.yaml (hunk 20 above)
Comment (mkykadir) [reply doc/config.example.yaml]: Implementing commit merged with #2825, ~can separate it to another PR~ actually that PR had documentation already

File: source/agora/node/Registry.d [hunk 21]
```diff
@@ -577,7 +580,8 @@ private SOA fromConfig (in ZoneConfig zone, Domain name, uint serial) @safe
 {
     return SOA(
         // mname, rname
-        Domain.fromString(zone.primary), Domain.fromString(zone.soa.email.value.replace('@', '.')),
+        Domain.fromString(zone.primary),
+        Domain.fromString(zone.soa.email.value.replace('@', '.')),
```
Comment (Geod24) [inline source/agora/node/Registry.d]: This is a style change. Either make it its own commit or revert it.

File: source/agora/node/Registry.d [hunk 22]
```diff
@@ -752,14 +766,23 @@ private struct ZoneData
...
+            if (this.config.soa.email.set)
+                this.type = ZoneType.primary;
+        }
+
+        static string serverType (ZoneType zone_type)
+        {
+            switch (zone_type)
+            {
+                case ZoneType.primary: return "primary (authoritative)";
+                default: return "Unknown";
```
Comment (Geod24) [inline source/agora/node/Registry.d]: I know this is fixed in a later commit, but we should not return "Unknown" here.

File: source/agora/node/Registry.d (hunk 11 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: `fromConfig` is module scope and I don't think it would be good to move it into `ZoneData`. Since `Serial` is not a configuration value, I will add a comment.

**Comment by mkykadir** [comment]
Rebased now depends on #2907

File: source/agora/node/Registry.d (hunk 18 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: I can either set default values for time values of `SOA` struct in `DNS` or keep this way of initializing it. `SOAConfig` already have default values which I wanted to utilize.
[This](https://github.com/mkykadir/agora/blob/a767a72dae19e96e6c423c70fe3223f99401de12/source/agora/node/Registry.d#L1026) is the reason why we need some default values.

File: source/agora/node/Config.d (hunk 15 above)
Comment (mkykadir) [reply source/agora/node/Config.d]: Currently, `primary` is for NS record which cannot cover secondary servers right now (thinking a way)
`primary_servers` are list of DNS servers for a secondary to perform AXFR from and recursion endpoint for cachings; if one is not reachable second one will be tried
`redirect_primary` must be an TCP/HTTP endpoint that has `postValidator/postFlashNode` available for a secondary server. Thus this cannot replace `primary` directly; but I am thinking for `primary` and `primary_servers` right now.

**Comment by mkykadir** [comment]
Rebased and resolved conversations except `primary` and `redirect_primary`

File: doc/config.example.yaml [hunk 23]
```diff
@@ -368,6 +369,26 @@ registry:
 
   # Zone responsible for flash nodes registration
   flash:
-    # Having separate authoritative server increases redundancy and reduces load,
+    # Having one or more secondary servers increase redundancy and reduces load,
     # hence this configuration, while unlikely in a small to medium network, is supported.
+    authoritative: true
+
+    # Redirect server for REST endpoints of the registry
+    redirect_primary: http://nsHidden.bosagora.io
```
Comment (Geod24) [inline doc/config.example.yaml]: Technically we don't have to redirect to a primary, so the name is a bit of a misnomer. But let's leave it as is for now.

File: source/agora/node/Registry.d [hunk 24]
```diff
@@ -834,31 +887,159 @@ private struct ZoneData
...
+                    format("DROP TABLE IF EXISTS registry_%s_utxo", zone_name));
+                this.db.execute(
+                    format("DROP TABLE IF EXISTS registry_%s_addresses", zone_name));
+            }
+
+            // DNS resolver is used to get SOA RR and performing AXFR
+            auto peer_addrs = this.config.primary_servers.map!(
+                peer => Address("dns://" ~ peer)
+            ).array();
+            this.resolver = node.makeDNSResolver(peer_addrs);
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Why doesn't the `Node` do that itself ?

File: source/agora/node/Registry.d [hunk 25]
```diff
@@ -834,31 +887,159 @@ private struct ZoneData
...
+            auto settings = new RestInterfaceSettings;
+            settings.baseURL = Address(this.config.redirect_primary);
+            settings.httpClientSettings = new HTTPClientSettings;
+            settings.httpClientSettings.connectTimeout = 2.seconds;
+            settings.httpClientSettings.readTimeout = 2.seconds;
+            this.redirect_primary = new RestInterfaceClient!NameRegistryAPI(settings);
+
+            this.expire_timer = this.taskman.createTimer(&this.disable);
+
+            this.soa_update_timer = this.taskman.setTimer(0.seconds, &this.updateSOA);
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Not going to work in unittests

**Review by Geod24 — COMMENTED** [review]
Second commit: Nope, we should not add Agora-specific definitions / value to `common.DNS`, especially with the provided reasoning. If we can't reach a primary, relying on semi arbitrary defaults seems prone to many issues.

File: source/agora/node/Registry.d (hunk 24 above)
Comment (mkykadir) [reply source/agora/node/Registry.d]: It's per-Zone. Each zone can have different list of primaries.

File: doc/config.example.yaml [hunk 26]
```diff
@@ -368,6 +369,26 @@ registry:
...
+
+    # List of servers to AXFR zone transfer from
+    #
+    # Servers are tried in the configuration order until zone transfer can be
+    # completed from a server.
+    primary_servers:
+      - ns1.bosagora.io
+
+    # A secondary server can be primary of an another server, thus secondary
```
Comment (Geod24) [inline doc/config.example.yaml]: No it can't. It can be an "upstream" server, but primary == ultimate authority on the zone.

File: doc/config.example.yaml (hunk 23 above)
Comment (mkykadir) [reply doc/config.example.yaml]: Renamed this and `primary_servers` and better documented

**Review by Geod24 — APPROVED** [review]
LGTM

**Comment by Geod24** [comment]
```
2022-01-27 09:28:19,226 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'unittest.bosagora.io.'
2022-01-27 09:28:19,226 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_realm_utxo in "SELECT * FROM registry_realm_utxo WHERE utxo = ?"
2022-01-27 09:28:19,226 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'validators.unittest.bosagora.io.'
2022-01-27 09:28:19,226 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_validator_utxo in "SELECT * FROM registry_validator_utxo WHERE utxo = ?"
2022-01-27 09:28:19,227 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'flash.unittest.bosagora.io.'
2022-01-27 09:28:19,227 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_flash_utxo in "SELECT * FROM registry_flash_utxo WHERE utxo = ?"
```

**Comment by Geod24** [comment]
The previous message logs were for one of the validator, but those ones are for the registry:
```
Registry logs:
Log for node: 10.8.8.8
======================================================================
2022-01-27 09:37:06,644 Info [agora.consensus.state.Ledger] - Last known block: #0 (0x8365f069fe37ee02f2c4dc6ad816702088fab5fc875c3c67b01f82c285aa2d90b605f57e068139eba1f20ce20578d712f75be4d8568c8f3a7a34604e72aa3175)
2022-01-27 09:37:06,647 Info [agora.consensus.state.Ledger] - Beginning externalization of block #0
2022-01-27 09:37:06,647 Info [agora.consensus.state.Ledger] - Transactions: 3 - Enrollments: 6
2022-01-27 09:37:06,647 Info [agora.consensus.state.Ledger] - Validators: Active: 0 - Signing: 000000 - Slashed: 0
2022-01-27 09:37:06,653 Info [agora.consensus.state.Ledger] - Completed externalization of block #0
2022-01-27 09:37:06,654 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'unittest.bosagora.io.'
2022-01-27 09:37:06,654 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_realm_utxo in "SELECT * FROM registry_realm_utxo WHERE utxo = ?"
2022-01-27 09:37:06,654 Info [agora.node.Registry] - Registry is primary (authoritative) DNS server for zone 'validators.unittest.bosagora.io.'
2022-01-27 09:37:06,654 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_validator_utxo in "SELECT * FROM registry_validator_utxo WHERE utxo = ?"
2022-01-27 09:37:06,655 Info [agora.node.Registry] - Registry is primary (authoritative) DNS server for zone 'flash.unittest.bosagora.io.'
2022-01-27 09:37:06,655 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_flash_utxo in "SELECT * FROM registry_flash_utxo WHERE utxo = ?"
```

Besides the same errors, the following sounds wrong:
> 2022-01-27 09:37:06,654 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'unittest.bosagora.io.'

**Comment by mkykadir** [comment]
> ```
> 2022-01-27 09:28:19,226 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'unittest.bosagora.io.'
> 2022-01-27 09:28:19,226 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_realm_utxo in "SELECT * FROM registry_realm_utxo WHERE utxo = ?"
> 2022-01-27 09:28:19,226 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'validators.unittest.bosagora.io.'
> 2022-01-27 09:28:19,226 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_validator_utxo in "SELECT * FROM registry_validator_utxo WHERE utxo = ?"
> 2022-01-27 09:28:19,227 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'flash.unittest.bosagora.io.'
> 2022-01-27 09:28:19,227 Error [agora.consensus.pool.Transaction] - SQLite error: (1) no such table: registry_flash_utxo in "SELECT * FROM registry_flash_utxo WHERE utxo = ?"
> ```

These logs are from #2822 ; I am checking for a table to understand if our configuration is changed from a primary to another

**Comment by mkykadir** [comment]
> > 2022-01-27 09:37:06,654 Info [agora.node.Registry] - Registry is caching (non-authoritative) DNS server for zone 'unittest.bosagora.io.'

Unittest config realm configured in [that way](https://github.com/bosagora/agora/blob/51a7ba60966c85f6bc315cf039dd5c380e1d64ce/source/agora/test/Base.d#L2073)

**Comment by Geod24** [comment]
Alright, we can fix this later. I've retriggered the CI to try to get a green run.

**Comment by Geod24** [comment]
MacOS passed.
